### PR TITLE
resume Bridge tasks

### PR DESCRIPTION
**Pull request by se3000**
Still need to handle some edge cases, but this is a rough outline.

The job runner now taking a RunResult might have some implications that we should talk through.

### CODE REVIEW

File: adapters/bridge.go [hunk 1]
```diff
@@ -19,38 +19,68 @@ type Bridge struct {
 
 // Perform sends a POST request containing the JSON of the input RunResult to
 // the external adapter specified in the BridgeType.
+// It records the RunResult reutrned to it, and optionally marks the RunResult pending.
```
Comment (dimroc) [inline adapters/bridge.go]: typo in reutrned

File: adapters/bridge.go [hunk 2]
```diff
@@ -19,38 +19,68 @@ type Bridge struct {
...
+//
+// If the Perform is resumed with a pending RunResult, the RunResult is marked
+// not pending and the RunResult is returned.
 func (ba *Bridge) Perform(input models.RunResult, _ *store.Store) models.RunResult {
-	in, err := json.Marshal(&input.Output)
+	if input.Pending {
+		return markNotPending(input)
```
Comment (dimroc) [inline adapters/bridge.go]: In this instance, don't we want to save the payload received from resuming to the run result?

File: web/job_runs_controller.go [hunk 3]
```diff
@@ -55,14 +55,40 @@ func (jrc *JobRunsController) Create(c *gin.Context) {
 	}
 }
 
+// Update marks the JobRun no longer pending, and resumes the Job's pipeline.
+func (jrc *JobRunsController) Update(c *gin.Context) {
+	id := c.Param("RunID")
```
Comment (dimroc) [inline web/job_runs_controller.go]: RunID and not id? This is all internal, based on the key you used in router, so it's not critical, but I would just stick with the convention of id. We're already in the JobRuns resource.

File: web/job_runs_controller.go [hunk 4]
```diff
@@ -55,14 +55,40 @@ func (jrc *JobRunsController) Create(c *gin.Context) {
...
+			"errors": []string{err.Error()},
+		})
+	} else if err := c.ShouldBindJSON(&rr); err != nil {
+		c.JSON(500, gin.H{
+			"errors": []string{err.Error()},
+		})
+	} else {
+		go func() {
+			if _, err = services.ExecuteRun(jr, jrc.App.Store, rr); err != nil {
+				logger.Errorw(fmt.Sprintf("Web initiator: %v", err.Error()))
```
Comment (dimroc) [inline web/job_runs_controller.go]: Let's place this in another method `resumeJob` for readability.

File: services/job_runner.go [hunk 5]
```diff
@@ -38,7 +38,7 @@ func BuildRun(job models.Job, store *store.Store) (models.JobRun, error) {
 // ExecuteRun starts the job and executes task runs within that job in the
 // order defined in the run for as long as they do not return errors. Results
 // are saved in the store (db).
-func ExecuteRun(run models.JobRun, store *store.Store, input models.JSON) (models.JobRun, error) {
+func ExecuteRun(run models.JobRun, store *store.Store, input models.RunResult) (models.JobRun, error) {
```
Comment (dimroc) [inline services/job_runner.go]: Why did we change the last parameter from `JSON` to `RunResult`? Was to ease the passing of the variable from the `JobRunsController` into `resumeJob`? Considering we only use `input.Data`, we should keep this as `JSON`.

File: adapters/bridge.go (hunk 2 above)
Comment (se3000) [reply adapters/bridge.go]: The payload received from resuming is now the `input`, so it is saved by the job runner once it is returned.

File: web/job_runs_controller.go (hunk 3 above)
Comment (se3000) [reply web/job_runs_controller.go]: Good catch, maybe from before I went with shallow routes.

File: services/job_runner.go (hunk 5 above)
Comment (se3000) [reply services/job_runner.go]: The main motivation behind that is that the external adapter may want to report an error. That's not yet covered in this PR, but is probably the biggest non-happy-path case I need to cover here. If the job runner takes a full RunResult then we don't have to come up with any funny ways of doing that or designating special keys in the RR's `Data`.

**Comment by se3000** [comment]
So this isn't 100% of the story, but I'm sending in a PR now to get some of these changes onto master sooner. The main piece missing is that it doesn't yet allow the external adapter to specify an error state for the run result. It can only move to completed.

There are also some edge cases in the controller that need to be covered and ideally tested a bit more:
- make sure that the run ID in the body is the same as the query params
- make sure this run result is currently pending
- test bad/malformed input scenarios
